keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1392 @@
+
+#### Phase1: Initialization (Responsibility of Contributors Implementing the KEP)
+
+1. Implement the test plan: [Validation Test Framework](https://docs.google.com/document/d/143KUaYcrEnxEWwPWrQH5P0MsF_V2Yrr8bsXJY76g86I/edit?resourcekey=0-K-n-yyITXtHbKVZnT9kAtw&tab=t.0#heading=h.hkbvgeat5xy8)
+2. Prototype and Initial API Selection (Core Team)
+*   Select a small set of representative Kubernetes API resources. (core/v1/replicationcontroller)
+*   Implement a working prototype by applying the entire process (adding IDL tags, generating code, updating tests) to these selected resources.
+3. Documentation and Contribution Guide (Core Team):
+*   Write documentation explaining how to:
+    *   Add validation tags (IDL tags).
+    *   Run validation-gen.
+    *   Update unit and E2E tests.
+*   Publish a contribution guide for the declarative validation migration.
+
+#### Phase2: Scaling the Migration (Responsibility of Contributors Implementing the KEP and broader community)
+
+1. Tracking Issue and Progress Management:
+*   Create a central tracking issue on GitHub.
+*   Break down the migration into smaller, manageable tasks. There are couple of options:
+    *   Per validation rule (recommended): Migrate a single validation rule for a specific field. E.g., `+k8s:minimum=0` for field `ReplicationControllerSpec.Replicas`
+    *   Per Field (recommended): Migrate all validation rules for a single field.
+    *   Per Type: Migrate all validation rules for a single API. e.g. all validation rules of `ReplicationControllerSpec`.` `Df
+    *   Per Group (not recommended): Migrate the entire API group/version.
+*   Label tasks appropriately.
+2. Community-Driven API Migration:
+*   Community:
+    *   Analyze existing handwritten validation.
+    *   Add appropriate IDL tags to API type definitions.
+    *   Run `validation-gen` to generate validation code.
+    *   Update unit tests to use the generated validation and ensure coverage.
+    *   Update E2E tests to verify behavior with declarative validation.
+    *   Submit pull requests (PRs) with the changes.
+*   Core Team:
+    *   Provide technical guidance and support to community contributors.
+    *   Review PRs.
+    *   Monitor the tracking issue and adjust the plan as needed.
+3. Using Schemas for Validation (Joint Effort):
+*   Core Team:
+    *   Enable validation through generated schemas for migrated resources (controlled by DeclarativeValidation feature gate).
+    *   Implement logic to populate default values from schemas.
+*   Community:
+    *   Run E2E tests with declarative validation enabled.
+
+
+#### 	Phase3: Finalization and GA (Core Team and community)
+
+
+
+1. After DeclarativeValidation reaches GA
+2. Deprecation of Legacy Validation is Announced
+3. ...Deprecation wait period passes (period adhering to community policy)…
+4. Legacy validation code that is being validated declaratively can safely be deleted
+
+### Tagging and Validating Against Internal vs Versioned Types
```

Comment:
wrap each of these blocks which need decisions with `<<[UNRESOLVED short context ]>>` and `<<[/UNRESOLVED]>>`

Reply:
Done